### PR TITLE
Bugfixing and grenade code cleanup

### LINKED ISSUE #4859: Toy grenades generate attack logs
Game ID: bUH-cNJD

### PR DESCRIPTION
**Pull request by Sindorman**
- Cleans up all grenade code by switching to absolute path for grenades procs. 

- Grenades now have `fake` var boolean that indicates if the grenade is fake

- Fake grenades now say in logs if they are fake. Fixes #4859

### CODE REVIEW

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 1]
```diff
@@ -15,176 +15,176 @@
...
+/obj/item/weapon/grenade/chem_grenade/Initialize()
+	. = ..()
+	var/datum/reagents/R = new/datum/reagents(1000)
+	reagents = R
+	R.my_atom = src
+
+/obj/item/weapon/grenade/chem_grenade/attack_self(mob/user as mob)
+	if(!stage || stage==1)
+		if(detonator)
 //				detonator.loc=src.loc
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: If we're cleaning up code, then delete this.

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 2]
```diff
@@ -15,176 +15,176 @@
...
-		if(istype(W,/obj/item/device/assembly_holder) && (!stage || stage==1) && path != 2)
-			var/obj/item/device/assembly_holder/det = W
-			if(istype(det.a_left,det.a_right.type) || (!isigniter(det.a_left) && !isigniter(det.a_right)))
-				user << "<span class='warning'>Assembly must contain one igniter.</span>"
-				return
-			if(!det.secured)
-				user << "<span class='warning'>Assembly must be secured with screwdriver.</span>"
-				return
+			detonator.detached()
+			usr.put_in_hands(detonator)
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: `usr` => `user`

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 3]
```diff
@@ -15,176 +15,176 @@
...
-				path = 1
-				if(beakers.len)
-					user << "<span class='notice'>You lock the assembly.</span>"
-					name = "grenade"
-				else
+			if(beakers.len)
+				user << "<span class='notice'>You lock the assembly.</span>"
+				name = "grenade"
+			else
 //					user << "<span class='warning'>You need to add at least one beaker before locking the assembly.</span>"
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: delet

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 4]
```diff
@@ -15,176 +15,176 @@
...
+	else if(is_type_in_list(W, allowed_containers) && (!stage || stage==1) && path != 2)
+		path = 1
+		if(beakers.len == 2)
+			user << "<span class='warning'>The grenade can not hold more containers.</span>"
+			return
+		else
+			if(W.reagents.total_volume)
+				user << "<span class='notice'>You add \the [W] to the assembly.</span>"
+				user.drop_item()
+				W.loc = src
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: `W.forceMove(src)`

File: code/game/objects/items/weapons/grenades/chem_grenade.dm [hunk 5]
```diff
@@ -15,176 +15,176 @@
...
-				src.reagents.touch(A)
+		for(var/atom/A in view(affected_area, src.loc))
+			if( A == src ) continue
+			src.reagents.touch(A)
 
-		if(istype(loc, /mob/living/carbon))		//drop dat grenade if it goes off in your hand
-			var/mob/living/carbon/C = loc
-			C.drop_from_inventory(src)
-			C.throw_mode_off()
+	if(istype(loc, /mob/living/carbon))		//drop dat grenade if it goes off in your hand
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/chem_grenade.dm]: `if (iscarbon(loc))`

File: code/game/objects/items/weapons/grenades/flashbang.dm [hunk 6]
```diff
@@ -5,120 +5,117 @@
...
-		..()
-		for(var/obj/structure/closet/L in hear(7, get_turf(src)))
-			if(locate(/mob/living/carbon/, L))
-				for(var/mob/living/carbon/M in L)
-					bang(get_turf(src), M)
-
+/obj/item/weapon/grenade/flashbang/prime()
+	..()
+	for(var/obj/structure/closet/L in hear(7, get_turf(src)))
+		if(locate(/mob/living/carbon/, L))
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/flashbang.dm]: If we're cleaning up. Remove this first `if`. Iterating over an empty list will serve the exact same purpose.

File: code/game/objects/items/weapons/grenades/flashbang.dm [hunk 7]
```diff
@@ -5,120 +5,117 @@
...
+	qdel(src)
+	return
 
-	proc/bang(var/turf/T , var/mob/living/carbon/M)					// Added a new proc called 'bang' that takes a location and a person to be banged.
-		if (locate(/obj/item/weapon/cloaking_device, M))			// Called during the loop that bangs people in lockers/containers and when banging
-			for(var/obj/item/weapon/cloaking_device/S in M)			// people in normal view.  Could theroetically be called during other explosions.
-				S.active = 0										// -- Polymorph
-				S.icon_state = "shield0"
+/obj/item/weapon/grenade/flashbang/proc/bang(var/turf/T , var/mob/living/carbon/M)  // Added a new proc called 'bang' that takes a location and a person to be banged.
+	if (locate(/obj/item/weapon/cloaking_device, M))								// Called during the loop that bangs people in lockers/containers and when banging
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/flashbang.dm]: Same deal here. Remove if check, replace with for.

File: code/game/objects/items/weapons/grenades/flashbang.dm [hunk 8]
```diff
@@ -5,120 +5,117 @@
...
-				if(HULK in M.mutations)
-					ear_safety += 1
-				if(istype(M:head, /obj/item/clothing/head/helmet))
-					ear_safety += 1
+	var/eye_safety = 0
+	var/ear_safety = 0
+	if(iscarbon(M))
+		eye_safety = M.eyecheck(TRUE)
+		if(ishuman(M))
+			if(istype(M:l_ear, /obj/item/clothing/ears/earmuffs) || istype(M:r_ear, /obj/item/clothing/ears/earmuffs))
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/flashbang.dm]: Remove `:` operator. Replace with `.`, typecast to stricter type if necessary.

File: code/game/objects/items/weapons/grenades/flashbang.dm [hunk 9]
```diff
@@ -5,120 +5,117 @@
...
+				ear_safety += 2
+			if(HULK in M.mutations)
+				ear_safety += 1
+			if(istype(M:head, /obj/item/clothing/head/helmet))
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/flashbang.dm]: Replace `:` with `.`.

File: code/game/objects/items/weapons/grenades/flashbang.dm [hunk 10]
```diff
@@ -5,120 +5,117 @@
...
+		flick("e_flash", M.flash)
+		M.Stun(2)
+		M.Weaken(10)
+			//Vaurca damage 15/01/16
+		var/mob/living/carbon/human/H = M
+		if(isvaurca(H))
+			var/obj/item/organ/eyes/E = H.get_eyes()
+			if(!E)
+				return
+			usr << span("alert", "Your eyes burn with the intense light of the flash!.")
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/flashbang.dm]: excuse, `usr`, here? Who is supposed to get this message, `H` I presume. Replace `usr` with `H`.

File: code/game/objects/items/weapons/grenades/grenade.dm [hunk 11]
```diff
@@ -11,6 +11,7 @@
 	slot_flags = SLOT_BELT
 	var/active = 0
 	var/det_time = 50
+	var/log = TRUE
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/grenade.dm]: This smells like a bad idea. Holocarp grenades have meaningful uses, as do fake grenades (in complaints, specifically). Could make a `var/fake` which interjects a `FAKE` into the attack message.

```dm
msg_admin_attack("[user.name] ([user.ckey]) primed \a [src] [fake ? "(FAKE) " : ""](<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[user.x];Y=[user.y];Z=[user.z]'>JMP</a>)",ckey=key_name(user))
```

File: code/game/objects/items/weapons/grenades/spawnergrenade.dm [hunk 12]
```diff
@@ -10,30 +10,30 @@
...
+		for(var/mob/living/carbon/human/M in viewers(T, null))
+			if(M.eyecheck(TRUE) < FLASH_PROTECTION_MODERATE)
+				flick("e_flash", M.flash)
+
+		for(var/i=1, i<=deliveryamt, i++)
+			var/atom/movable/x = new spawner_type
+			if(newvars && length(newvars))
+				for(var/v in newvars)
+					x.vars[v] = newvars[v]
+			x.loc = T
```
Comment (skull132) [inline code/game/objects/items/weapons/grenades/spawnergrenade.dm]: `x.forceMove(T)`

**Comment by Sindorman** [comment]
@skull132 I mean, it fixes bug of fake grenades showing logs(which is not a bug). Yeah, we could change log message about fake grenade, I think it is more if admins want to see fake grenades. You want me to target development?

**Comment by skull132** [comment]
@Sindorman ye target develop. And ye keep the logs, just append a clear indication that they're fake into it.

**Comment by Sindorman** [comment]
@skull132 Changed. Review please.

**Pull request merged**